### DISCUSSION TITLE
Github package **public** npm registry

### DISCUSSION THREAD

**Discussion by Jym77** [renovatebot/renovate#26475] [Request Help] [answered]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### What is your question?

 How to configure renovate to access **public** npm packages on Github registry?

MNWE: [renovate-public-mnwe dashboard](https://github.com/Jym77/renovate-public-mnwe/issues/2)

I have some **public** npm package stored on the Github registry: [alfa-rules](https://github.com/Siteimprove/alfa/pkgs/npm/alfa-rules), which I use from another repository: [renovate-public-mnwe](https://github.com/Jym77/renovate-public-mnwe), with an outdated version.
I am hoping that Renovate can automagically access version and create PRs, as usual.

However, I am getting a "Renovate failed to look up the following dependencies: `Failed to look up npm package @siteimprove/alfa-rules`." error in [Renovate dashboard Issue](https://github.com/Jym77/renovate-public-mnwe/issues/2). I assume this is due to the lack of token for accessing the Github registry.

There is quite a lot of [documentation on accessing **private** packages on Github npm registry](https://docs.renovatebot.com/getting-started/private-packages/#npm), but I cannot find anything specific to public packages (including in discussions, looks like they are mostly concerned with private packages).

As far as I understand Github npm registry, a token must be provided to access it. However, in the case of public packages, any token (with `packages:read`) is enough, most notably the `GITHUB_TOKEN` (when running a workflow). As illustrated with [my build workflow](https://github.com/Jym77/renovate-public-mnwe/blob/main/.github/workflows/integrate.yml), which is [running `yarn install` smoothly](https://github.com/Jym77/renovate-public-mnwe/actions/runs/7384390761/job/20087239440) simply adding a `yarn config set npmScopes.siteimprove.npmAuthToken ${{ secrets.GITHUB_TOKEN }}` to access Github npm registry with the `GITHUB_TOKEN`.

* I could create my own PAT, pre-encrypt it, and add it to Renovate config. This seems a bit overkill. (plus, require maintenance for rolling tokens, …; which is quite some overhead in enterprise setting with strict security rules, even if the token only has `packages:read` permission…) (plus, given that it will also have access to private packages, this dosen't sound super good in terms of "least privilege")
* It looks like self-hosted Renovate can access the `GITHUB_TOKEN` of the action running it. Which would be pretty good. But I somewhat need to run it with the Renovate App (again, due to enterprise policies).
* I tried (unsuccessfully) to access the `GITHUB_TOKEN` from `renovate.json` but failed. Not even sure this is possible…
* Am I just missing a preset that is already doing that trick (not seeing any, the [:githubComToken(<`arg0`>) one](https://docs.renovatebot.com/presets-default/#githubcomtokenarg0) also requires a pre-encrypted token)?
* The logs seems to contain some `hostRules` with tokens, but not for `https://npm.pkg.github.com/`.

So, do I need to create my own PAT for this use case, or is there a way to tell Renovate "just use any `packages:read` token"?

### Logs (if relevant)

<details><summary>Logs</summary>

Config, containing some `hostRules`:
```
DEBUG: Combined config
{
  "config": {
    "token": "***********",
    "privateKey": "***********",
    "privateKeyOld": "***********",
    "hostRules": [
      {
        "hostType": "docker",
        "matchHost": "docker.io",
        "username": "mdpprodrenovate",
        "password": "***********"
      },
      {
        "hostType": "docker",
        "matchHost": "hub.docker.com",
        "username": "mdpprodrenovatedc2",
        "password": "***********"
      },
      {
        "hostType": "github",
        "matchHost": "api.github.com",
        "token": "***********"
      },
      {
        "hostType": "github-tags",
        "matchHost": "api.github.com",
        "token": "***********"
      },
      {
        "hostType": "github-releases",
        "matchHost": "api.github.com",
        "token": "***********"
      },
      {
        "hostType": "pod",
        "matchHost": "api.github.com",
        "token": "***********"
      },
      {
        "hostType": "terraform-provider",
        "matchHost": "api.github.com",
        "token": "***********"
      },
      {
        "hostType": "rubygems",
        "matchHost": "github.com",
        "token": "***********"
      }
    ],
    "extends": [
      "mergeConfidence:all-badges"
    ],
    "prFooter": "This PR has been generated by [Mend Renovate](https://www.mend.io/free-developer-tools/renovate/). View repository job log [here](https://developer.mend.io/{{platform}}/{{repository}}).",
    "prHeader": "[![Mend Renovate](https://app.renovatebot.com/images/banner.svg)](https://renovatebot.com)",
    "redisUrl": "redis://mend-developer-platform-renovate-prod.aqffol.ng.0001.use1.cache.amazonaws.com:6379",
    "binarySource": "install",
    "platformCommit": true,
    "repositoryCache": "enabled",
    "cacheTtlOverride": {
      "datasource-docker-tags": 90,
      "datasource-docker-digest": 90,
      "datasource-docker-labels": 90,
      "datasource-docker-releases": 90
    },
    "onboardingConfig": {
      "$schema": "https://docs.renovatebot.com/renovate-schema.json",
      "extends": [
        "config:recommended"
      ]
    },
    "gitIgnoredAuthors": [
      "29139614+renovate[bot]@users.noreply.github.com"
    ],
    "customizeDashboard": {
      "repoProblemsHeader": "These problems occurred while renovating this repository. [View logs](https://developer.mend.io//{{platform}}/{{repository}})."
    },
    "repositoryCacheType": "s3://mend-developer-platform-prod/renovate/",
    "dependencyDashboardFooter": "\n- [ ] <!-- manual job -->Check this box to trigger a request for Renovate to run again on this repository\n",
    "allowedPostUpgradeCommands": [
      "^git add --all$",
      "^git reset$",
      "^npx beachball change( --no-fetch)? --no-commit --type (patch|none) --message '{{{commitMessage}}}'$",
      "^pwd$"
    ],
    "allowPostUpgradeCommandTemplating": true,
    "onboardingNoDeps": true,
    "onboarding": true,
    "forkProcessing": "enabled",
    "requireConfig": "required",
    "platform": "github",
    "username": "renovate[bot]",
    "repositories": [
      "Jym77/renovate-public-mnwe"
    ],
    "gitAuthor": "renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>"
  }
}
```

Running into problem:
```
DEBUG: hostRules: no authentication for npm.pkg.github.com
DEBUG: GET https://npm.pkg.github.com/@siteimprove%2Falfa-rules = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=401 retryCount=0, duration=46)
DEBUG: Failed to look up npm package @siteimprove/alfa-rules
{
  "dependency": "@siteimprove/alfa-rules"
  "packageFile": "package.json"
}

(…)

 "npm": [
      {
        "deps": [
          {
            "depType": "dependencies",
            "depName": "@siteimprove/alfa-rules",
            "currentValue": "^0.70.0",
            "datasource": "npm",
            "prettyDepType": "dependency",
            "registryUrls": [
              "https://npm.pkg.github.com"
            ],
            "lockedVersion": "0.70.0",
            "updates": [],
            "packageName": "@siteimprove/alfa-rules",
            "versioning": "npm",
            "warnings": [
              {
                "topic": "@siteimprove/alfa-rules",
                "message": "Failed to look up npm package @siteimprove/alfa-rules"
              }
            ]
          },
```

</details>

**Comment by rarkins** [comment, marked as answer]
For now you need to add a PAT for this. We previously tried reusing the app token and had problems so reverted it because we weren't sure why the problems occurred. Of course we'd like it to "just work" using the same token which renovate already has, assuming it has sufficient privileges.

**Reply by Jym77** [reply]
OK. Thanks for the answer.

**Comment by evan-boone** [comment]
Is it possible to turn on this feature experimentally when used the hosted app? It seems like the doc still references it - https://docs.renovatebot.com/modules/platform/github/#package-registry-credentials. If so, what exactly do you need to add to a renovate.json to turn this feature on?  

And are there any plans to continue working on this feature? If so, what is the timeline?

**Reply by rarkins** [reply]
It's only possible to set env variables using self-hosted

**Reply by stekern** [reply]
@rarkins Are there any plans to either allow users of the hosted Renovate app to opt-in for the experiment, or look into making the necessary changes to make this feature generally available?

**Reply by rarkins** [reply]
No to the hosted app, because it didn't work. And we're happy to make the changes once someone can work out what went wrong.

**Reply by stekern** [reply]
Gotcha. I was under the impression that GitHub had introduced support for using GitHub App tokens tokens with the GitHub Packages API, but it seems I was wrong. GitHub today only supports authentication with GitHub Packages using either a classic personal access token or a GitHub Actions token.

You're blocked by https://github.com/orgs/community/discussions/26920#discussioncomment-3253960 (and perhaps also https://github.com/github/roadmap/issues/558).

**Reply by viceice** [reply]
@rarkins @HonkingGoose I think we can extend the warning inside the docs with those new information.

**Reply by rarkins** [reply]
@stekern thanks for tracking down that info!

**Reply by HonkingGoose** [reply]
I'm trying to fix the docs problem with this PR:

- #26697
